### PR TITLE
*: implementation of grpc drivers

### PR DESCRIPTION
**Pull request by mcuadros**
This is the implementation of the grpc driver, it's working, but it a preliminary version. I should work in logging (server and from driver), cleanup of the code and error handling (mainly for the old drivers)

PTAL to the architecture, the code may contain leftovers and/or quick and dirty stuff.

### CODE REVIEW

File: driver.go [hunk 1]
```diff
@@ -1,81 +1,142 @@
...
+	Start() error
+	Stop() error
+	Service() protocol.ProtocolServiceClient
+}
+
+type DriverInstance struct {
+	ctx  context.Context
+	p    *runtime.Process
+	c    runtime.Container
+	i    runtime.DriverImage
```
Comment (abeaumont) [inline driver.go]: `proc` and `img` may be better names than `p` and `i`.

File: driver.go [hunk 2]
```diff
@@ -1,81 +1,142 @@
...
+	if err := i.c.Start(); err != nil {
+		logrus.Errorf("error starting container %s (%s): %s", i.i.Name(), i.c.ID(), err)
+		return err
+	}
+
+	if err := i.dial(); err != nil {
+		return err
+	}
+
+	if err := i.loadVersion(); err != nil {
```
Comment (abeaumont) [inline driver.go]: What is this for?

Comment (mcuadros) [reply driver.go]: is a validation of the connection.

File: driver.go (hunk 1 above)
Comment (mcuadros) [reply driver.go]: ok

**Comment by mcuadros** [comment]
done @juanjux @abeaumont

File: Makefile [hunk 3]
```diff
@@ -88,7 +90,8 @@ $(NOVENDOR_PACKAGES):
 	$(GO_GET) $@
 
 $(VENDOR_PATH):
-	$(GLIDE) install
+	$(GLIDE) install; \
+	rm -rf vendor/github.com/Sirupsen/;
```
Comment (abeaumont) [inline Makefile]: Why is this needed?

File: cmd/bblfshd/main.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+	overrides := make(map[string]string)
+	for _, img := range strings.Split(os.Getenv("BBLFSH_DRIVER_IMAGES"), ";") {
+		if img = strings.TrimSpace(img); img == "" {
+			continue
+		}
+
+		fields := strings.Split(img, "=")
+		if len(fields) != 2 {
+			logger.Errorf("invalid image driver format %s", img)
+			os.Exit(1)
```
Comment (abeaumont) [inline cmd/bblfshd/main.go]: Having so many exit points makes code hard to reuse/test. It may be better to just return the error and let the caller handle it properly.

File: glide.lock [hunk 5]
```diff
@@ -166,10 +174,10 @@ imports:
   version: 7af7a1e09ba336d2ea14b1ce73bf693c6837dbf6
 - name: github.com/seccomp/libseccomp-golang
   version: f6ec81daf48e41bf48b475afc7fe06a26bfb72d1
-- name: github.com/Sirupsen/logrus
-  version: f006c2ac4710855cf0f916dd6b77acf6b048dc6e
 - name: github.com/sirupsen/logrus
   version: 89742aefa4b206dcf400792f3bd35b542998eb3b
+- name: github.com/Sirupsen/logrus
```
Comment (abeaumont) [inline glide.lock]: Do we need to have these duplicated?

File: glide.yaml [hunk 6]
```diff
@@ -37,7 +37,8 @@ import:
 - package: google.golang.org/grpc
   version: ^1.4.2
 - package: gopkg.in/bblfsh/sdk.v1
-  version: ^1.0.0-rc1
+  version: 83b3b64de663069fd7e0a12d08db4c00da3c17ce
```
Comment (abeaumont) [inline glide.yaml]: This should be reverted back to a named version once SDK changes are released.

File: cmd/bblfshd/main.go (hunk 4 above)
Comment (mcuadros) [reply cmd/bblfshd/main.go]: Witch caller, is a main

File: Makefile (hunk 3 above)
Comment (mcuadros) [reply Makefile]: https://github.com/sirupsen/logrus/issues/570#issuecomment-313933276
Will be deleted soon.

File: glide.lock (hunk 5 above)
Comment (mcuadros) [reply glide.lock]: Yep, sadly https://github.com/sirupsen/logrus/issues/570#issuecomment-313933276 and some projects are using the other one. My next PR will be updating all the dependencies.

File: cmd/bblfshd/main.go [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	logger server.Logger
+)
+
+func init() {
+	cmd := flag.NewFlagSet("server", flag.ExitOnError)
+	network = cmd.String("network", "tcp", "network type: tcp, tcp4, tcp6, unix or unixpacket.")
+	address = cmd.String("address", "localhost:9432", "address to listen.")
+	storage = cmd.String("storage", "/var/lib/bblfshd", "path where all the runtime information is stored.")
```
Comment (juanjux) [inline cmd/bblfshd/main.go]: It would be nice to have an option to not cache the images; it would be very useful when you're debugging interactions between drivers and the servers and you change the local image frequently. But that's outside the scope of this PR and could be done later.

File: cmd/bblfshd/main.go (hunk 4 above)
Comment (abeaumont) [reply cmd/bblfshd/main.go]: `main` function is the caller. I guess this is a matter of taste, since it's a direct call from main.

Comment (mcuadros) [reply cmd/bblfshd/main.go]: I just wanted to make the main function simple and avoid error handling there...

File: cmd/bblfshd/main.go (hunk 7 above)
Comment (mcuadros) [reply cmd/bblfshd/main.go]: The images should be stored and updated if need it. A feature to be develop is a command line to to force this update in the server, as `docker pull <image>`

File: runtime/container_test.go [hunk 8]
```diff
@@ -231,7 +232,7 @@ func TestContainerStartWaitExit1(t *testing.T) {
 		Stdout: out,
 	}
 
-	c, err := rt.Container(d, p)
+	c, err := rt.Container("start-wait-exit1", d, p, nil)
```
Comment (juanjux) [inline runtime/container_test.go]: exit1?

File: runtime/runtime.go [hunk 9]
```diff
@@ -66,9 +67,15 @@ func (r *Runtime) ListDrivers() ([]*DriverImageStatus, error) {
 }
 
 // Container returns a container for the given DriverImage and Process
-func (r *Runtime) Container(d DriverImage, p *Process) (Container, error) {
+func (r *Runtime) Container(id string, d DriverImage, p *Process, f ConfigFactory) (Container, error) {
+	//id := NewULID().String()
```
Comment (juanjux) [inline runtime/runtime.go]: Remove?

File: daemon/daemon.go [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+	transport := s.Transport
+	if transport == "" {
+		transport = defaultTransport
+	}
+
+	ref := fmt.Sprintf("bblfsh/%s-driver:latest", lang)
+	switch transport {
+	case "docker":
+		ref = "//" + ref
+	}
```
[the remark this answers is not in the document]
Comment (mcuadros) [reply daemon/daemon.go]: this is old code, and planned to be deprecated.

File: daemon/daemon.go [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+
+func getDriverInstanceOptions(logger server.Logger) *Options {
+	opts := &Options{}
+
+	var l *logrus.Logger
+	switch i := logger.(type) {
+	case *logrus.Logger:
+		l = i
+	case *logrus.Entry:
+		l = i.Logger
```
[the remark this answers is not in the document]
Comment (mcuadros) [reply daemon/daemon.go]: I was thinking about this, I could find nothing better, without mixing concepts.

File: runtime/container_test.go (hunk 8 above)
Comment (mcuadros) [reply runtime/container_test.go]: its the name of the test :/

File: runtime/runtime.go (hunk 9 above)
Comment (mcuadros) [reply runtime/runtime.go]: sure

File: runtime/container_test.go (hunk 8 above)
Comment (juanjux) [reply runtime/container_test.go]: Ahhh, didn't notice that.

File: cmd/bblfshd/main.go [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+		overrides[lang] = image
+	}
+
+	return overrides
+}
+
+func handleGracefullyShutdown(l net.Listener, d *daemon.Daemon) {
+	var gracefulStop = make(chan os.Signal)
+	signal.Notify(gracefulStop, syscall.SIGTERM)
+	signal.Notify(gracefulStop, syscall.SIGINT)
```
[the remark this answers is not in the document]
Comment (mcuadros) [reply cmd/bblfshd/main.go]: You have many signal to catch like SIGTERM, SIGINT, SIGQUIT, SIGABRT, SIGKILL, but what I said are more or less the more commons, I prefer keep does, and see if they some others a missing.

[the remark this answers is not in the document]
Comment (juanjux) [reply cmd/bblfshd/main.go]: Only Jesus can catch SIGKILL :) other than that I'm Ok with handling these two.
